sys/kern/cred_syscalls.c

```diff
@@ -0,0 +1,87 @@
+#include <sys/cred.h>
+#include <sys/proc.h>
+#include <sys/errno.h>
+
+int do_getresuid(proc_t *p, uid_t *ruid, uid_t *euid, uid_t *suid) {
+  cred_t cr = p->p_cred;
+  if (ruid != NULL)
+    *ruid = cr.cr_ruid;
+  if (euid != NULL)
+    *euid = cr.cr_euid;
+  if (suid != NULL)
+    *suid = cr.cr_suid;
+  return 0;
+}
+
+int do_getresgid(proc_t *p, gid_t *rgid, gid_t *egid, gid_t *sgid) {
+  cred_t cr = p->p_cred;
+  if (rgid != NULL)
+    *rgid = cr.cr_rgid;
+  if (egid != NULL)
+    *egid = cr.cr_egid;
+  if (sgid != NULL)
+    *sgid = cr.cr_sgid;
+  return 0;
+}
+
+int do_setresuid(proc_t *p, uid_t ruid, uid_t euid, uid_t suid) {
+  proc_lock(p);
+
+  cred_t *oldcred = &p->p_cred;
+  cred_t newcred; /* XXX: maybe not on stack */
+  cred_copy(&newcred, oldcred);
```

Comment:
I think that you are wrong here.

`void cred_copy(cred_t *src, cred_t *dest)`
So you copy _random_ value from the stack to your old credentials and lost them forever.

Reply:
Thanks. I changed `cred_copy` to look like `memcpy`. Now it will be less confusing. 
```void cred_copy(cred_t *dest, cred_t src)```